From bf19558c26fb865bcd573920cd1978a40f4ffb9c Mon Sep 17 00:00:00 2001
From: Sergey Chupligin <s.chupligin@omprussia.ru>
Date: Thu, 4 Apr 2019 14:13:49 +0300
Subject: [PATCH] [systemd] Fix build with glibc-2.28

Since glibc-2.27 memfd_create implemented by glibc
Since glibc-2.26 makedev macros moving to sys/sysmacros

---
 src/basic/btrfs-util.c                              | 1 +
 src/basic/hashmap.c                                 | 1 +
 src/basic/missing.h                                 | 9 +++++++--
 src/basic/terminal-util.c                           | 5 +++--
 src/basic/util.c                                    | 9 +++++----
 src/boot/bootctl.c                                  | 1 +
 src/core/automount.c                                | 1 +
 src/core/cgroup.c                                   | 1 +
 src/core/device.c                                   | 1 +
 src/core/service.c                                  | 1 +
 src/core/umount.c                                   | 1 +
 src/cryptsetup/cryptsetup.c                         | 1 +
 src/fsck/fsck.c                                     | 1 +
 src/gpt-auto-generator/gpt-auto-generator.c         | 1 +
 src/hibernate-resume/hibernate-resume.c             | 1 +
 src/journal/journalctl.c                            | 1 +
 src/libsystemd/sd-bus/bus-dump.c                    | 2 ++
 src/libsystemd/sd-bus/bus-kernel.c                  | 1 +
 src/libsystemd/sd-device/device-enumerator.c        | 1 +
 src/libsystemd/sd-device/device-private.c           | 1 +
 src/libsystemd/sd-device/sd-device.c                | 1 +
 src/libudev/libudev-device.c                        | 1 +
 src/login/logind-session-dbus.c                     | 1 +
 src/login/logind-session-device.c                   | 1 +
 src/nspawn/nspawn.c                                 | 1 +
 src/systemctl/systemctl.c                           | 2 +-
 src/test/test-libudev.c                             | 1 +
 src/tmpfiles/tmpfiles.c                             | 1 +
 src/tty-ask-password-agent/tty-ask-password-agent.c | 1 +
 src/udev/cdrom_id/cdrom_id.c                        | 1 +
 src/udev/scsi_id/scsi_serial.c                      | 1 +
 src/udev/udev-event.c                               | 1 +
 src/udev/udev-node.c                                | 1 +
 src/udev/udev-rules.c                               | 1 +
 src/udev/udevadm-info.c                             | 1 +
 src/udev/udevd.c                                    | 1 +
 36 files changed, 49 insertions(+), 9 deletions(-)

diff --git a/src/basic/btrfs-util.c b/src/basic/btrfs-util.c
index 074deeccd..cab5f9a3d 100644
--- a/src/basic/btrfs-util.c
+++ b/src/basic/btrfs-util.c
@@ -22,6 +22,7 @@
 #include <stdlib.h>
 #include <sys/vfs.h>
 #include <sys/stat.h>
+#include <sys/sysmacros.h>
 
 #ifdef HAVE_LINUX_BTRFS_H
 #include <linux/btrfs.h>
diff --git a/src/basic/hashmap.c b/src/basic/hashmap.c
index 7d2a4160c..5882cd5d7 100644
--- a/src/basic/hashmap.c
+++ b/src/basic/hashmap.c
@@ -20,6 +20,7 @@
   along with systemd; If not, see <http://www.gnu.org/licenses/>.
 ***/
 
+#include <sys/sysmacros.h>
 #include <stdlib.h>
 #include <errno.h>
 #include <pthread.h>
diff --git a/src/basic/missing.h b/src/basic/missing.h
index 34ab0254d..2bbb4ab94 100644
--- a/src/basic/missing.h
+++ b/src/basic/missing.h
@@ -132,6 +132,7 @@ static inline int pivot_root(const char *new_root, const char *put_old) {
 }
 #endif
 
+
 #ifndef __NR_memfd_create
 #  if defined __x86_64__
 #    define __NR_memfd_create 319
@@ -158,9 +159,11 @@ static inline int pivot_root(const char *new_root, const char *put_old) {
 #endif
 
 #ifndef HAVE_MEMFD_CREATE
-static inline int memfd_create(const char *name, unsigned int flags) {
+static inline int missing_memfd_create(const char *name, unsigned int flags) {
         return syscall(__NR_memfd_create, name, flags);
 }
+
+#  define memfd_create missing_memfd_create
 #endif
 
 #ifndef __NR_getrandom
@@ -1009,9 +1012,11 @@ static inline pid_t raw_getpid(void) {
 #  endif
 #endif
 
-static inline int renameat2(int oldfd, const char *oldname, int newfd, const char *newname, unsigned flags) {
+static inline int missing_renameat2(int oldfd, const char *oldname, int newfd, const char *newname, unsigned flags) {
         return syscall(__NR_renameat2, oldfd, oldname, newfd, newname, flags);
 }
+
+#  define renameat2 missing_renameat2
 #endif
 
 #ifndef RENAME_NOREPLACE
diff --git a/src/basic/terminal-util.c b/src/basic/terminal-util.c
index cf55263bb..dd4b55dce 100644
--- a/src/basic/terminal-util.c
+++ b/src/basic/terminal-util.c
@@ -20,6 +20,7 @@
 #include <sys/ioctl.h>
 #include <sys/types.h>
 #include <sys/stat.h>
+#include <sys/sysmacros.h>
 #include <termios.h>
 #include <unistd.h>
 #include <fcntl.h>
@@ -42,7 +43,7 @@ static volatile unsigned cached_columns = 0;
 static volatile unsigned cached_lines = 0;
 
 int chvt(int vt) {
-        _cleanup_close_ int fd;
+        _cleanup_close_ int fd = -1;
 
         fd = open_terminal("/dev/tty0", O_RDWR|O_NOCTTY|O_CLOEXEC|O_NONBLOCK);
         if (fd < 0)
@@ -529,7 +530,7 @@ int terminal_vhangup_fd(int fd) {
 }
 
 int terminal_vhangup(const char *name) {
-        _cleanup_close_ int fd;
+        _cleanup_close_ int fd = -1;
 
         fd = open_terminal(name, O_RDWR|O_NOCTTY|O_CLOEXEC|O_NONBLOCK);
         if (fd < 0)
diff --git a/src/basic/util.c b/src/basic/util.c
index f752595ca..8f42e3152 100644
--- a/src/basic/util.c
+++ b/src/basic/util.c
@@ -57,6 +57,7 @@
 #include <sys/xattr.h>
 #include <sys/statvfs.h>
 #include <sys/file.h>
+#include <sys/sysmacros.h>
 #include <linux/fs.h>
 
 /* When we include libgen.h because we need dirname() we immediately
@@ -2316,7 +2317,7 @@ bool is_device_path(const char *path) {
 }
 
 int dir_is_empty(const char *path) {
-        _cleanup_closedir_ DIR *d;
+        _cleanup_closedir_ DIR *d = NULL;
 
         d = opendir(path);
         if (!d)
@@ -2649,7 +2650,7 @@ char *ellipsize(const char *s, size_t length, unsigned percent) {
 }
 
 int touch_file(const char *path, bool parents, usec_t stamp, uid_t uid, gid_t gid, mode_t mode) {
-        _cleanup_close_ int fd;
+        _cleanup_close_ int fd = -1;
         int r;
 
         assert(path);
@@ -2862,7 +2863,7 @@ static int do_execute(char **directories, usec_t timeout, char *argv[]) {
                 return log_oom();
 
         STRV_FOREACH(directory, directories) {
-                _cleanup_closedir_ DIR *d;
+                _cleanup_closedir_ DIR *d = NULL;
                 struct dirent *de;
 
                 d = opendir(*directory);
@@ -5122,7 +5123,7 @@ int getpeersec(int fd, char **ret) {
 
 /* This is much like like mkostemp() but is subject to umask(). */
 int mkostemp_safe(char *pattern, int flags) {
-        _cleanup_umask_ mode_t u;
+        _cleanup_umask_ mode_t u = NULL;
         int fd;
 
         assert(pattern);
diff --git a/src/boot/bootctl.c b/src/boot/bootctl.c
index ac1711b31..9b3f5e53c 100644
--- a/src/boot/bootctl.c
+++ b/src/boot/bootctl.c
@@ -26,6 +26,7 @@
 #include <assert.h>
 #include <sys/statfs.h>
 #include <sys/stat.h>
+#include <sys/sysmacros.h>
 #include <errno.h>
 #include <string.h>
 #include <unistd.h>
diff --git a/src/core/automount.c b/src/core/automount.c
index 4af381b4b..f3c1335f9 100644
--- a/src/core/automount.c
+++ b/src/core/automount.c
@@ -26,6 +26,7 @@
 #include <fcntl.h>
 #include <sys/epoll.h>
 #include <sys/stat.h>
+#include <sys/sysmacros.h>
 #include <linux/auto_fs4.h>
 #include <linux/auto_dev-ioctl.h>
 
diff --git a/src/core/cgroup.c b/src/core/cgroup.c
index 6474e08bd..89a275e5d 100644
--- a/src/core/cgroup.c
+++ b/src/core/cgroup.c
@@ -19,6 +19,7 @@
   along with systemd; If not, see <http://www.gnu.org/licenses/>.
 ***/
 
+#include <sys/sysmacros.h>
 #include <fcntl.h>
 #include <fnmatch.h>
 
diff --git a/src/core/device.c b/src/core/device.c
index e7efcf0f0..a7dbe6b6b 100644
--- a/src/core/device.c
+++ b/src/core/device.c
@@ -21,6 +21,7 @@
 
 #include <errno.h>
 #include <sys/epoll.h>
+#include <sys/sysmacros.h>
 #include <libudev.h>
 
 #include "log.h"
diff --git a/src/core/service.c b/src/core/service.c
index b790ec98b..fc6ebc1dc 100644
--- a/src/core/service.c
+++ b/src/core/service.c
@@ -19,6 +19,7 @@
   along with systemd; If not, see <http://www.gnu.org/licenses/>.
 ***/
 
+#include <sys/sysmacros.h>
 #include <errno.h>
 #include <signal.h>
 #include <unistd.h>
diff --git a/src/core/umount.c b/src/core/umount.c
index d59b5d0ff..fdfdbbd66 100644
--- a/src/core/umount.c
+++ b/src/core/umount.c
@@ -24,6 +24,7 @@
 #include <string.h>
 #include <sys/mount.h>
 #include <sys/swap.h>
+#include <sys/sysmacros.h>
 #include <linux/loop.h>
 #include <linux/dm-ioctl.h>
 
diff --git a/src/cryptsetup/cryptsetup.c b/src/cryptsetup/cryptsetup.c
index 74fa90a23..24fd383d6 100644
--- a/src/cryptsetup/cryptsetup.c
+++ b/src/cryptsetup/cryptsetup.c
@@ -22,6 +22,7 @@
 #include <string.h>
 #include <errno.h>
 #include <sys/mman.h>
+#include <sys/sysmacros.h>
 #include <mntent.h>
 
 #include <libcryptsetup.h>
diff --git a/src/fsck/fsck.c b/src/fsck/fsck.c
index bd3051f30..5db1f5e39 100644
--- a/src/fsck/fsck.c
+++ b/src/fsck/fsck.c
@@ -28,6 +28,7 @@
 #include <sys/file.h>
 #include <sys/stat.h>
 #include <sys/prctl.h>
+#include <sys/sysmacros.h>
 
 #include "sd-bus.h"
 #include "sd-device.h"
diff --git a/src/gpt-auto-generator/gpt-auto-generator.c b/src/gpt-auto-generator/gpt-auto-generator.c
index 0a34f86be..036eb23c5 100644
--- a/src/gpt-auto-generator/gpt-auto-generator.c
+++ b/src/gpt-auto-generator/gpt-auto-generator.c
@@ -22,6 +22,7 @@
 #include <unistd.h>
 #include <stdlib.h>
 #include <sys/statfs.h>
+#include <sys/sysmacros.h>
 #include <blkid/blkid.h>
 
 #include "sd-id128.h"
diff --git a/src/hibernate-resume/hibernate-resume.c b/src/hibernate-resume/hibernate-resume.c
index 1f3b16990..abc3564e5 100644
--- a/src/hibernate-resume/hibernate-resume.c
+++ b/src/hibernate-resume/hibernate-resume.c
@@ -22,6 +22,7 @@
 #include <stdio.h>
 #include <errno.h>
 #include <sys/stat.h>
+#include <sys/sysmacros.h>
 
 #include "log.h"
 #include "util.h"
diff --git a/src/journal/journalctl.c b/src/journal/journalctl.c
index ce2dd9da2..f3a8099aa 100644
--- a/src/journal/journalctl.c
+++ b/src/journal/journalctl.c
@@ -33,6 +33,7 @@
 #include <poll.h>
 #include <sys/stat.h>
 #include <sys/inotify.h>
+#include <sys/sysmacros.h>
 #include <linux/fs.h>
 
 #include "sd-journal.h"
diff --git a/src/libsystemd/sd-bus/bus-dump.c b/src/libsystemd/sd-bus/bus-dump.c
index a6b05eb88..d370a8516 100644
--- a/src/libsystemd/sd-bus/bus-dump.c
+++ b/src/libsystemd/sd-bus/bus-dump.c
@@ -19,6 +19,8 @@
   along with systemd; If not, see <http://www.gnu.org/licenses/>.
 ***/
 
+#include <sys/sysmacros.h>
+
 #include "util.h"
 #include "capability.h"
 #include "strv.h"
diff --git a/src/libsystemd/sd-bus/bus-kernel.c b/src/libsystemd/sd-bus/bus-kernel.c
index 577a8b44c..73840a17c 100644
--- a/src/libsystemd/sd-bus/bus-kernel.c
+++ b/src/libsystemd/sd-bus/bus-kernel.c
@@ -27,6 +27,7 @@
 #include <malloc.h>
 #include <sys/mman.h>
 #include <sys/prctl.h>
+#include <sys/sysmacros.h>
 
 /* When we include libgen.h because we need dirname() we immediately
  * undefine basename() since libgen.h defines it as a macro to the POSIX
diff --git a/src/libsystemd/sd-device/device-enumerator.c b/src/libsystemd/sd-device/device-enumerator.c
index 5eb37e16c..4622d52f0 100644
--- a/src/libsystemd/sd-device/device-enumerator.c
+++ b/src/libsystemd/sd-device/device-enumerator.c
@@ -18,6 +18,7 @@
   along with systemd; If not, see <http://www.gnu.org/licenses/>.
 ***/
 
+#include <sys/sysmacros.h>
 #include "util.h"
 #include "prioq.h"
 #include "strv.h"
diff --git a/src/libsystemd/sd-device/device-private.c b/src/libsystemd/sd-device/device-private.c
index 0ec966774..ef9226b43 100644
--- a/src/libsystemd/sd-device/device-private.c
+++ b/src/libsystemd/sd-device/device-private.c
@@ -20,6 +20,7 @@
 
 #include <ctype.h>
 #include <sys/types.h>
+#include <sys/sysmacros.h>
 #include <net/if.h>
 
 #include "util.h"
diff --git a/src/libsystemd/sd-device/sd-device.c b/src/libsystemd/sd-device/sd-device.c
index 7cea5a074..1fc792435 100644
--- a/src/libsystemd/sd-device/sd-device.c
+++ b/src/libsystemd/sd-device/sd-device.c
@@ -20,6 +20,7 @@
 
 #include <ctype.h>
 #include <sys/types.h>
+#include <sys/sysmacros.h>
 #include <net/if.h>
 
 #include "util.h"
diff --git a/src/libudev/libudev-device.c b/src/libudev/libudev-device.c
index 9a8d68210..3032b0574 100644
--- a/src/libudev/libudev-device.c
+++ b/src/libudev/libudev-device.c
@@ -32,6 +32,7 @@
 #include <sys/stat.h>
 #include <sys/ioctl.h>
 #include <sys/socket.h>
+#include <sys/sysmacros.h>
 #include <linux/sockios.h>
 
 #include "sd-device.h"
diff --git a/src/login/logind-session-dbus.c b/src/login/logind-session-dbus.c
index 563153e2d..c9226b025 100644
--- a/src/login/logind-session-dbus.c
+++ b/src/login/logind-session-dbus.c
@@ -21,6 +21,7 @@
 
 #include <errno.h>
 #include <string.h>
+#include <sys/sysmacros.h>
 
 #include "util.h"
 #include "strv.h"
diff --git a/src/login/logind-session-device.c b/src/login/logind-session-device.c
index 656f268db..32b94ec70 100644
--- a/src/login/logind-session-device.c
+++ b/src/login/logind-session-device.c
@@ -25,6 +25,7 @@
 #include <string.h>
 #include <sys/ioctl.h>
 #include <sys/types.h>
+#include <sys/sysmacros.h>
 
 #include "util.h"
 #include "missing.h"
diff --git a/src/nspawn/nspawn.c b/src/nspawn/nspawn.c
index 837947ee2..44e33f83c 100644
--- a/src/nspawn/nspawn.c
+++ b/src/nspawn/nspawn.c
@@ -39,6 +39,7 @@
 #include <sys/personality.h>
 #include <linux/loop.h>
 #include <sys/file.h>
+#include <sys/sysmacros.h>
 
 #ifdef HAVE_SELINUX
 #include <selinux/selinux.h>
diff --git a/src/systemctl/systemctl.c b/src/systemctl/systemctl.c
index 3cb5f6186..55fcd8a23 100644
--- a/src/systemctl/systemctl.c
+++ b/src/systemctl/systemctl.c
@@ -4611,7 +4611,7 @@ static int init_home_and_lookup_paths(char **user_home, char **user_runtime, Loo
 }
 
 static int cat_file(const char *filename, bool newline) {
-        _cleanup_close_ int fd;
+        _cleanup_close_ int fd = -1;
 
         fd = open(filename, O_RDONLY|O_CLOEXEC|O_NOCTTY);
         if (fd < 0)
diff --git a/src/test/test-libudev.c b/src/test/test-libudev.c
index 34c49b969..fa7cc14bb 100644
--- a/src/test/test-libudev.c
+++ b/src/test/test-libudev.c
@@ -22,6 +22,7 @@
 #include <unistd.h>
 #include <getopt.h>
 #include <sys/epoll.h>
+#include <sys/sysmacros.h>
 
 #include "libudev.h"
 #include "udev-util.h"
diff --git a/src/tmpfiles/tmpfiles.c b/src/tmpfiles/tmpfiles.c
index 8f29256c6..0dfd3f209 100644
--- a/src/tmpfiles/tmpfiles.c
+++ b/src/tmpfiles/tmpfiles.c
@@ -36,6 +36,7 @@
 #include <fnmatch.h>
 #include <sys/stat.h>
 #include <sys/xattr.h>
+#include <sys/sysmacros.h>
 #include <linux/fs.h>
 
 #include "log.h"
diff --git a/src/tty-ask-password-agent/tty-ask-password-agent.c b/src/tty-ask-password-agent/tty-ask-password-agent.c
index 82cbf95f1..5f30eb5ad 100644
--- a/src/tty-ask-password-agent/tty-ask-password-agent.c
+++ b/src/tty-ask-password-agent/tty-ask-password-agent.c
@@ -30,6 +30,7 @@
 #include <unistd.h>
 #include <getopt.h>
 #include <sys/signalfd.h>
+#include <sys/sysmacros.h>
 #include <fcntl.h>
 
 #include "util.h"
diff --git a/src/udev/cdrom_id/cdrom_id.c b/src/udev/cdrom_id/cdrom_id.c
index 775da0355..a8becfb52 100644
--- a/src/udev/cdrom_id/cdrom_id.c
+++ b/src/udev/cdrom_id/cdrom_id.c
@@ -32,6 +32,7 @@
 #include <sys/stat.h>
 #include <sys/time.h>
 #include <sys/ioctl.h>
+#include <sys/sysmacros.h>
 #include <linux/cdrom.h>
 
 #include "libudev.h"
diff --git a/src/udev/scsi_id/scsi_serial.c b/src/udev/scsi_id/scsi_serial.c
index 3691a69d4..ffc2e5e97 100644
--- a/src/udev/scsi_id/scsi_serial.c
+++ b/src/udev/scsi_id/scsi_serial.c
@@ -20,6 +20,7 @@
 #include <sys/types.h>
 #include <sys/ioctl.h>
 #include <sys/stat.h>
+#include <sys/sysmacros.h>
 #include <stdio.h>
 #include <errno.h>
 #include <string.h>
diff --git a/src/udev/udev-event.c b/src/udev/udev-event.c
index 476122278..a6f678097 100644
--- a/src/udev/udev-event.c
+++ b/src/udev/udev-event.c
@@ -29,6 +29,7 @@
 #include <sys/epoll.h>
 #include <sys/wait.h>
 #include <sys/signalfd.h>
+#include <sys/sysmacros.h>
 
 #include "netlink-util.h"
 #include "event-util.h"
diff --git a/src/udev/udev-node.c b/src/udev/udev-node.c
index e730fb45f..2d2560faf 100644
--- a/src/udev/udev-node.c
+++ b/src/udev/udev-node.c
@@ -24,6 +24,7 @@
 #include <errno.h>
 #include <dirent.h>
 #include <sys/stat.h>
+#include <sys/sysmacros.h>
 
 #include "udev.h"
 #include "smack-util.h"
diff --git a/src/udev/udev-rules.c b/src/udev/udev-rules.c
index 43255fb9d..2d9a5ee3c 100644
--- a/src/udev/udev-rules.c
+++ b/src/udev/udev-rules.c
@@ -28,6 +28,7 @@
 #include <dirent.h>
 #include <fnmatch.h>
 #include <time.h>
+#include <sys/sysmacros.h>
 
 #include "udev.h"
 #include "path-util.h"
diff --git a/src/udev/udevadm-info.c b/src/udev/udevadm-info.c
index b3d5565c4..438220141 100644
--- a/src/udev/udevadm-info.c
+++ b/src/udev/udevadm-info.c
@@ -25,6 +25,7 @@
 #include <getopt.h>
 #include <fcntl.h>
 #include <sys/stat.h>
+#include <sys/sysmacros.h>
 
 #include "udev.h"
 #include "udev-util.h"
diff --git a/src/udev/udevd.c b/src/udev/udevd.c
index 28ac44fb8..dc706d302 100644
--- a/src/udev/udevd.c
+++ b/src/udev/udevd.c
@@ -39,6 +39,7 @@
 #include <sys/stat.h>
 #include <sys/ioctl.h>
 #include <sys/inotify.h>
+#include <sys/sysmacros.h>
 
 #include "sd-daemon.h"
 #include "sd-event.h"
-- 
2.19.1

